File: Heap/MaxHeap.java
package Heap;

import java.util.Arrays;

public class MaxHeap {
    private int[] heap;
    private int size;
    private int capacity;

    public MaxHeap(int capacity) {
        this.capacity = capacity;
        this.size = 0;
        this.heap = new int[capacity];
    }

    private int parent(int index) {
        return (index - 1) / 2;
    }

    private int leftChild(int index) {
        return 2 * index + 1;
    }

    private int rightChild(int index) {
        return 2 * index + 2;
    }

    private void swap(int index1, int index2) {
        int temp = heap[index1];
        heap[index1] = heap[index2];
        heap[index2] = temp;
    }

    public void insert(int key) {
        if (size == capacity) {
            heap = Arrays.copyOf(heap, capacity * 2);
            capacity *= 2;
        }
        heap[size] = key;
        int current = size;
        size++;

        while (current != 0 && heap[parent(current)] < heap[current]) {
            swap(current, parent(current));
            current = parent(current);
        }
    }

    public int extractMax() {
        if (size == 0) {
            throw new IllegalStateException("Heap is empty");
        }
        int root = heap[0];
        heap[0] = heap[size - 1];
        size--;
        heapifyDown(0);
        return root;
    }

    private void heapifyDown(int index) {
        int largest = index;
        int left = leftChild(index);
        int right = rightChild(index);

        if (left < size && heap[left] > heap[largest]) {
            largest = left;
        }
        if (right < size && heap[right] > heap[largest]) {
            largest = right;
        }
        if (largest != index) {
            swap(index, largest);
            heapifyDown(largest);
        }
    }

    public int getMax() {
        if (size == 0) {
            throw new IllegalStateException("Heap is empty");
        }
        return heap[0];
    }

    public static void main(String[] args) {
        MaxHeap maxHeap = new MaxHeap(10);
        maxHeap.insert(3);
        maxHeap.insert(2);
        maxHeap.insert(1);
        maxHeap.insert(7);
        maxHeap.insert(8);
        maxHeap.insert(4);
        maxHeap.insert(10);
        maxHeap.insert(16);
        maxHeap.insert(12);

        System.out.println("Max value: " + maxHeap.getMax());
        System.out.println("Extracted max value: " + maxHeap.extractMax());
        System.out.println("Max value after extraction: " + maxHeap.getMax());
    }
}

File: Heap/maxHeap.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *heap;
    int size;
    int capacity;
} MaxHeap;

MaxHeap* createMaxHeap(int capacity) {
    MaxHeap* maxHeap = (MaxHeap*)malloc(sizeof(MaxHeap));
    maxHeap->capacity = capacity;
    maxHeap->size = 0;
    maxHeap->heap = (int*)malloc(capacity * sizeof(int));
    return maxHeap;
}

int parent(int index) {
    return (index - 1) / 2;
}

int leftChild(int index) {
    return 2 * index + 1;
}

int rightChild(int index) {
    return 2 * index + 2;
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void insertMaxHeap(MaxHeap* maxHeap, int key) {
    if (maxHeap->size == maxHeap->capacity) {
        maxHeap->capacity *= 2;
        maxHeap->heap = (int*)realloc(maxHeap->heap, maxHeap->capacity * sizeof(int));
    }
    maxHeap->heap[maxHeap->size] = key;
    int current = maxHeap->size;
    maxHeap->size++;

    while (current != 0 && maxHeap->heap[parent(current)] < maxHeap->heap[current]) {
        swap(&maxHeap->heap[current], &maxHeap->heap[parent(current)]);
        current = parent(current);
    }
}

void heapifyDownMaxHeap(MaxHeap* maxHeap, int index) {
    int largest = index;
    int left = leftChild(index);
    int right = rightChild(index);

    if (left < maxHeap->size && maxHeap->heap[left] > maxHeap->heap[largest]) {
        largest = left;
    }
    if (right < maxHeap->size && maxHeap->heap[right] > maxHeap->heap[largest]) {
        largest = right;
    }
    if (largest != index) {
        swap(&maxHeap->heap[index], &maxHeap->heap[largest]);
        heapifyDownMaxHeap(maxHeap, largest);
    }
}

int extractMax(MaxHeap* maxHeap) {
    if (maxHeap->size == 0) {
        printf("Heap is empty\n");
        return -1;
    }
    int root = maxHeap->heap[0];
    maxHeap->heap[0] = maxHeap->heap[maxHeap->size - 1];
    maxHeap->size--;
    heapifyDownMaxHeap(maxHeap, 0);
    return root;
}

int getMax(MaxHeap* maxHeap) {
    if (maxHeap->size == 0) {
        printf("Heap is empty\n");
        return -1;
    }
    return maxHeap->heap[0];
}

int main() {
    MaxHeap* maxHeap = createMaxHeap(10);
    insertMaxHeap(maxHeap, 3);
    insertMaxHeap(maxHeap, 2);
    insertMaxHeap(maxHeap, 1);
    insertMaxHeap(maxHeap, 7);
    insertMaxHeap(maxHeap, 8);
    insertMaxHeap(maxHeap, 4);
    insertMaxHeap(maxHeap, 10);
    insertMaxHeap(maxHeap, 16);
    insertMaxHeap(maxHeap, 12);

    printf("Max value: %d\n", getMax(maxHeap));
    printf("Extracted max value: %d\n", extractMax(maxHeap));
    printf("Max value after extraction: %d\n", getMax(maxHeap));

    free(maxHeap->heap);
    free(maxHeap);
    return 0;
}